### PR TITLE
Here it is: the PR to merge the error handling staging branch into develop

### PR DESCRIPTION
**Pull request by balos1**
🙌

### CODE REVIEW

**Comment by balos1** [comment]
There are 6 commits at the very end of this PR (after the #391 merge commit) that fixed bugs that prevented the CI from passing. You can click on each of those commits to see those additional changes.

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:486]:
```suggestion
  message("Found clang-tidy: ${CLANG_TIDY_PATH}")
```

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:487]: Move below of IWYU block below
```suggestion
```

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:511]:
```suggestion
  make_directory(${SUNDIALS_DEV_IWYU_DIR}) 
  set(CMAKE_C_INCLUDE_WHAT_YOU_USE ${IWYU_PATH}
```

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:489]: Should this be added as an advanced option?

File: scripts/iwyu.imp
Comment (gardner48) [inline scripts/iwyu.imp:14]: Add new line at end of file

**Review by gardner48 — CHANGES_REQUESTED** [review]

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:511]: The variable `SUNDIALS_DEV_IWYU_DIR` is unset

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:505]:
```suggestion
set(SUNDIALS_DEV_IWYU_DIR ${CMAKE_BINARY_DIR}/iwyu/)

if(SUNDIALS_DEV_IWYU)
```

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:510]:
```suggestion
  message(STATUS "Found IWYU: ${IWYU_PATH}")
```

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:505]: Actually, is `SUNDIALS_DEV_IWYU_DIR` needed? IWYU doesn't seem to have an option to write its suggestions to a file.

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [reply cmake/SundialsSetupCompilers.cmake:511]: Maybe this isn't needed, see comment below.

**Review by drreynolds — DISMISSED** [review]
I'll defer to @gardner48 for the CMake, GitHub actions, "iwyu", tidy and format changes.  From what else I see, this looks great.

File: cmake/SundialsSetupCompilers.cmake
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake:483]: There is no reason it should need to be changed, so I think its best to leave it out for now to reduce the number of options that display for developers.

File: cmake/SundialsSetupCompilers.cmake
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake:499]: As you have noted, it is not needed.

File: cmake/SundialsSetupCompilers.cmake
Comment (gardner48) [inline cmake/SundialsSetupCompilers.cmake:126]: Will this flag cause any problems on Windows or compilers other than gcc/clang?

**Review by gardner48 — COMMENTED** [review]
One question about the warning flag, otherwise this looks good to me

**Review by gardner48 — APPROVED** [review]

File: cmake/SundialsSetupCompilers.cmake
Comment (balos1) [reply cmake/SundialsSetupCompilers.cmake]: As per our offline discussions, this would cause a problem so we removed it.

**Pull request merged**
